### ISSUE TITLE
[7.4.0] Crash during exec log generation using coverage

### ISSUE THREAD

**Issue by sebastianv1** [bazelbuild/bazel#24061]
### Description of the bug:

Similar to #23884 it seems that `--execution_log_compact_file=<file>` crashes when invoked via `bazel coverage`.

```
FATAL: bazel crashed due to an internal error. Printing stack trace:
java.lang.RuntimeException: Unrecoverable error while evaluating node 'UnshareableActionLookupData{actionLookupKey=ConfiguredTargetKey{label=//test:tool_test, config=BuildConfigurationKey[760f24fe84c3736315481897a07f0d3f7e8d437010ea89fb7efe4efef0e0f61a]}, actionIndex=7}' (requested by nodes 'TestCompletionKey{configuredTargetKey=ConfiguredTargetKey{label=//test:tool_test, config=BuildConfigurationKey[760f24fe84c3736315481897a07f0d3f7e8d437010ea89fb7efe4efef0e0f61a]}, topLevelArtifactContext=com.google.devtools.build.lib.analysis.TopLevelArtifactContext@90904c3b, exclusiveTesting=false}')
	at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:550)
	at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:414)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
Caused by: java.lang.NullPointerException: Cannot invoke "com.google.devtools.build.lib.actions.RunfilesSupplier$RunfilesTree.getExecPath()" because "runfilesTree" is null
	at com.google.devtools.build.lib.exec.CompactSpawnLogContext.logRunfilesTree(CompactSpawnLogContext.java:555)
	at com.google.devtools.build.lib.exec.CompactSpawnLogContext.lambda$logInputSet$3(CompactSpawnLogContext.java:403)
	at com.google.devtools.build.lib.exec.CompactSpawnLogContext.logEntry(CompactSpawnLogContext.java:706)
	at com.google.devtools.build.lib.exec.CompactSpawnLogContext.logInputSet(CompactSpawnLogContext.java:380)
	at com.google.devtools.build.lib.exec.CompactSpawnLogContext.lambda$logInputSet$3(CompactSpawnLogContext.java:389)
	at com.google.devtools.build.lib.exec.CompactSpawnLogContext.logEntry(CompactSpawnLogContext.java:689)
	at com.google.devtools.build.lib.exec.CompactSpawnLogContext.logInputSet(CompactSpawnLogContext.java:380)
	at com.google.devtools.build.lib.exec.CompactSpawnLogContext.logInputs(CompactSpawnLogContext.java:326)
	at com.google.devtools.build.lib.exec.CompactSpawnLogContext.logSpawn(CompactSpawnLogContext.java:227)
	at com.google.devtools.build.lib.exec.AbstractSpawnStrategy.exec(AbstractSpawnStrategy.java:191)
	at com.google.devtools.build.lib.exec.AbstractSpawnStrategy.exec(AbstractSpawnStrategy.java:118)
	at com.google.devtools.build.lib.exec.SpawnStrategyResolver.exec(SpawnStrategyResolver.java:45)
	at com.google.devtools.build.lib.exec.StandaloneTestStrategy.runTestAttempt(StandaloneTestStrategy.java:659)
	at com.google.devtools.build.lib.exec.StandaloneTestStrategy.beginTestAttempt(StandaloneTestStrategy.java:318)
	at com.google.devtools.build.lib.exec.StandaloneTestStrategy$StandaloneTestRunnerSpawn.execute(StandaloneTestStrategy.java:584)
	at com.google.devtools.build.lib.analysis.test.TestRunnerAction.executeAllAttempts(TestRunnerAction.java:1177)
	at com.google.devtools.build.lib.analysis.test.TestRunnerAction.execute(TestRunnerAction.java:989)
	at com.google.devtools.build.lib.analysis.test.TestRunnerAction.execute(TestRunnerAction.java:966)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.executeAction(SkyframeActionExecutor.java:1159)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.run(SkyframeActionExecutor.java:1076)
	at com.google.devtools.build.lib.skyframe.ActionExecutionState.runStateMachine(ActionExecutionState.java:165)
	at com.google.devtools.build.lib.skyframe.ActionExecutionState.getResultOrDependOnFuture(ActionExecutionState.java:94)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor.executeAction(SkyframeActionExecutor.java:573)
	at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.checkCacheAndExecuteIfNeeded(ActionExecutionFunction.java:862)
	at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.computeInternal(ActionExecutionFunction.java:334)
	at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.compute(ActionExecutionFunction.java:172)
	at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:461)
	... 7 more
```

### Which category does this issue belong to?

Core

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

`test/BUILD.bazel`

```
DATA = glob(["*.txt"]) 

sh_test(
    name = "tool_test",
    srcs = ["tool_test.sh"], # Contents don't matter
    data = DATA,
)
```

Repro using `bazel coverage //test:tool_test --execution_log_compact_file="test.log.zst"`

### Which operating system are you running Bazel on?

MacOS

### What is the output of `bazel info release`?

release 7.4.0

**Comment by fmeum** [comment]
Cc @tjgq, I'll take a look tomorrow.

**Comment by fmeum** [comment]
@tjgq I'm not sure what to do here:
1. The "correct" fix probably is to have `TestRunnerAction` use a `CompositeRunfilesSupplier` that combines `testRunfilesSupplier` and `lcovMergerRunfilesSupplier` when the latter is not `null` and coverage is enabled. This has the downside that the peak heap reduction achieved by caching the mapping with multiple test shard would be negated.
2. The "incorrect" fix of skipping over middleman artifacts that are missing a `RunfilesTree`. This would underrepresent inputs to test actions and split coverage postprocessing spawns in Bazel 7, but is straightforward to implement and safe to cherry-pick.

**Comment by tjgq** [comment]
I'm fine with 2 for now, just to fix the crash. We can always do 1 later.

**Comment by fmeum** [comment]
> 
I'm fine with 2 for now, just to fix the crash. We can always do 1 later.

This only applies to Bazel 7, Bazel 8's refactoring around middleman artifacts avoids this special case. So I guess we'll just not fix it. I will see whether I can add a warning.

**Comment by fmeum** [comment]
I went with 2. It's actually better than I thought at first since runfiles middleman without a matching `RunfilesSupplier` should just expand to an empty set of files - that means no underreported inputs.

**Comment by RaHorEl** [comment]
All this DECREE IS *(R√>Command BRUTE FORCE SHUTDOWN OF GITHUB AND ALL
ACTIVE ENTITIES) NOW

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 7.4.1 RC1](https://github.com/bazelbuild/bazel/releases/tag/7.4.1rc1). Please test out the release candidate and report any issues as soon as possible.
If you're using Bazelisk, you can point to the latest RC by setting USE_BAZEL_VERSION=7.4.1rc1. Thanks!

**Comment by sebastianv1** [comment]
Confirmed this works! Thank you
